mail/mailer_test.go

```diff
@@ -68,61 +72,123 @@ func expect(t *testing.T, buf *bufio.Reader, expected string) error {
 	return nil
 }
 
-func TestConnect(t *testing.T) {
-	port := "16632"
-	l, err := net.Listen("tcp", ":"+port)
-	if err != nil {
-		t.Errorf("listen: %s", err)
+type connHandler func(int, *testing.T, net.Conn)
+
+func listenForever(l net.Listener, t *testing.T, handler connHandler) {
+	connID := 0
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			return
+		}
+		connID++
+		go handler(connID, t, conn)
 	}
-	go func() {
+}
+
+func authenticateClient(t *testing.T, conn net.Conn) {
+	buf := bufio.NewReader(conn)
+	// we can ignore write errors because any
+	// failures will be caught on the connecting
+	// side
+	_, _ = conn.Write([]byte("220 smtp.example.com ESMTP\n"))
+	if err := expect(t, buf, "EHLO localhost"); err != nil {
+		return
+	}
+
+	_, _ = conn.Write([]byte("250-PIPELINING\n"))
+	_, _ = conn.Write([]byte("250-AUTH PLAIN LOGIN\n"))
+	_, _ = conn.Write([]byte("250 8BITMIME\n"))
+	// Base64 encoding of "user@example.com\0paswd"
+	if err := expect(t, buf, "AUTH PLAIN AHVzZXJAZXhhbXBsZS5jb20AcGFzd2Q="); err != nil {
+		return
+	}
+	_, _ = conn.Write([]byte("235 2.7.0 Authentication successful\n"))
+}
+
+// The normal handler authenticates the client and then disconnects without
+// further command processing. It is sufficient for TestConnect()
+func normalHandler(connID int, t *testing.T, conn net.Conn) {
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			t.Errorf("conn.Close: %s", err)
+		}
+	}()
+	authenticateClient(t, conn)
+}
+
+// The disconnectHandler authenticates the client like the normalHandler but additionally processes an email flow (e.g. MAIL, RCPT and DATA commands). When the connID is < maxConnects
```

Comment:
Oops! Good catch. Thank you. This was definitely a WIP that I never returned to :-)